File: src/commands/clearstarterproject.js
import { access, readdir, rm } from "node:fs/promises";
import { spawn } from "node:child_process";
import path from "node:path";

const output = process.stdout;
const color = output.isTTY && !process.env.NO_COLOR;
const bold = (s) => (color ? `\x1b[1m${s}\x1b[0m` : s);
const dim = (s) => (color ? `\x1b[2m${s}\x1b[0m` : s);
const cyan = (s) => (color ? `\x1b[36m${s}\x1b[0m` : s);
const green = (s) => (color ? `\x1b[32m${s}\x1b[0m` : s);

// Starter scaffold written by `eco startproject` (createCompositionScaffold).
// Each entry removes the placeholder runtime files but keeps the service
// contract files (.gitignore, .env.example) so a replacement can move in.
const STARTER_PATHS = [
  { path: "frontend/package.json", label: "frontend package.json" },
  { path: "frontend/index.js", label: "frontend starter server" },
  { path: "frontend/index.html", label: "frontend starter page" },
  { path: "frontend/images", label: "frontend starter images", dir: true },
  { path: "backend/Cargo.toml", label: "backend Cargo.toml" },
  { path: "backend/src", label: "backend starter source", dir: true }
];

async function pathExists(targetPath) {
  try {
    await access(targetPath);
    return true;
  } catch {
    return false;
  }
}

function findCompositionArg(args) {
  const positionals = args.filter((arg) => !arg.startsWith("--"));
  return positionals[0] || null;
}

async function findCompositionDir(cwd) {
  // Walk up from cwd; a `<project>_composition` directory is the composition repo.
  let dir = path.resolve(cwd);
  const { root } = path.parse(dir);
  while (true) {
    try {
      const entries = await readdir(dir, { withFileTypes: true });
      const composition = entries.find(
        (entry) => entry.isDirectory() && entry.name.endsWith("_composition")
      );
      if (composition) {
        return path.join(dir, composition.name);
      }
    } catch {}
    if (dir === root) break;
    dir = path.dirname(dir);
  }
  return null;
}

async function gitStatusClean(cwd) {
  return new Promise((resolve) => {
    const child = spawn("git", ["status", "--porcelain"], { cwd, stdio: ["ignore", "pipe", "ignore"] });
    let stdout = "";
    child.stdout.on("data", (c) => { stdout += c; });
    child.on("error", () => resolve(null));
    child.on("exit", (code) => {
      if (code !== 0) return resolve(null);
      const lines = stdout.trim().split("\n").filter(Boolean);
      // Ignore the untouched README.md-style untracked docs; flag tracked changes
      // only so an in-progress real app is never silently wiped.
      const tracked = lines.filter((line) => line.startsWith(" M ") || line.startsWith("M ") || line.startsWith(" D ") || line.startsWith("D "));
      resolve(tracked.length === 0);
    });
  });
}

function clearstarterprojectHelp() {
  output.write(`eco clearstarterproject [path]\n\nRemove the placeholder starter runtime files from a <project>_composition\nrepository, leaving the service contract (.gitignore, .env.example) and the\nrepository itself intact so a real frontend/backend can replace the starter.\n\nArguments:\n  path   path to the composition repository (default: nearest *_composition dir)\n\nOptions:\n  --commit   commit the removal in the composition repository\n  --dry-run  list what would be removed without deleting\n`);
}

export async function runClearStarterProject(args) {
  if (args.includes("help") || args.includes("--help") || args.includes("-h")) {
    clearstarterprojectHelp();
    return;
  }

  const dryRun = args.includes("--dry-run");
  const commit = args.includes("--commit");
  const explicitPath = findCompositionArg(args);

  const compositionDir = explicitPath
    ? path.resolve(explicitPath)
    : await findCompositionDir(process.cwd());

  if (!compositionDir) {
    throw new Error("No <project>_composition directory found. Pass the path or run inside the estate.");
  }
  if (!(await pathExists(compositionDir))) {
    throw new Error(`Composition directory not found: ${compositionDir}`);
  }

  const isRepo = await pathExists(path.join(compositionDir, ".git"));
  if (isRepo && !(await gitStatusClean(compositionDir)) && !dryRun) {
    throw new Error(
      `${compositionDir} has uncommitted changes to tracked files. ` +
      "Commit or stash them before clearing the starter project."
    );
  }

  const present = [];
  for (const entry of STARTER_PATHS) {
    if (await pathExists(path.join(compositionDir, entry.path))) {
      present.push(entry);
    }
  }

  output.write(`\n${bold("Eco starter project")}  ${dim(compositionDir)}\n`);
  if (present.length === 0) {
    output.write(`  ${green("Nothing to clear")} — no starter files present.\n\n`);
    return;
  }

  for (const entry of present) {
    output.write(`  ${dryRun ? "would remove" : "removing"}  ${cyan(entry.path)}\n`);
  }

  if (dryRun) {
    output.write("\n");
    return;
  }

  for (const entry of present) {
    await rm(path.join(compositionDir, entry.path), { recursive: true, force: true });
  }

  // Drop empty frontend/backend trees so the repo reflects "no app yet".
  for (const serviceDir of ["frontend", "backend"]) {
    const full = path.join(compositionDir, serviceDir);
    if (await pathExists(full)) {
      const remaining = await readdir(full);
      if (remaining.length === 0) {
        await rm(full, { recursive: true, force: true });
        output.write(`  removing empty   ${cyan(serviceDir + "/")}\n`);
      }
    }
  }

  if (commit && isRepo) {
    await new Promise((resolve, reject) => {
      const child = spawn("git", ["add", "-A"], { cwd: compositionDir, stdio: "inherit" });
      child.on("error", reject);
      child.on("exit", (code) => code === 0 ? resolve() : reject(new Error(`git add exited with code ${code}`)));
    });
    await new Promise((resolve, reject) => {
      const child = spawn("git", ["commit", "-m", "clear: remove starter project scaffold"], { cwd: compositionDir, stdio: "inherit" });
      child.on("error", reject);
      child.on("exit", (code) => code === 0 ? resolve() : reject(new Error(`git commit exited with code ${code}`)));
    });
    output.write(`\n  ${green("Committed")} removal in ${compositionDir}\n`);
  } else if (isRepo && !commit) {
    output.write(`\n  ${dim("Commit the removal with: git -C <composition> commit -m \"clear: remove starter scaffold\"")}\n`);
  }

  output.write("\n");
}

File: rust/src/commands/clearstarterproject.rs
use crate::util;
use std::path::{Path, PathBuf};

const STARTER_PATHS: &[(&str, &str, bool)] = &[
    ("frontend/package.json", "frontend package.json", false),
    ("frontend/index.js", "frontend starter server", false),
    ("frontend/index.html", "frontend starter page", false),
    ("frontend/images", "frontend starter images", true),
    ("backend/Cargo.toml", "backend Cargo.toml", false),
    ("backend/src", "backend starter source", true),
];

fn path_exists(p: &Path) -> bool {
    p.exists()
}

fn find_composition_arg(args: &[String]) -> Option<String> {
    args.iter().find(|a| !a.starts_with("--")).cloned()
}

fn find_composition_dir(cwd: &Path) -> Option<PathBuf> {
    let mut dir = cwd.to_path_buf();
    loop {
        if let Ok(entries) = std::fs::read_dir(&dir) {
            for entry in entries.flatten() {
                if entry.path().is_dir() {
                    if let Some(name) = entry.file_name().to_str() {
                        if name.ends_with("_composition") {
                            return Some(entry.path());
                        }
                    }
                }
            }
        }
        let parent = dir.parent()?.to_path_buf();
        if parent == dir {
            break;
        }
        dir = parent;
    }
    None
}

fn git_status_clean(cwd: &Path) -> Option<bool> {
    let result = util::run_capture(
        "git",
        &["status".to_string(), "--porcelain".to_string()],
        cwd,
    )
    .ok()?;
    if result.code != 0 {
        return None;
    }
    let lines: Vec<String> = result
        .stdout
        .split('\n')
        .map(|s| s.to_string())
        .filter(|s| !s.trim().is_empty())
        .collect();
    let tracked = lines
        .iter()
        .filter(|l| {
            l.starts_with(" M ") || l.starts_with("M ") || l.starts_with(" D ") || l.starts_with("D ")
        })
        .count();
    Some(tracked == 0)
}

fn help_text() {
    let text = r#"eco clearstarterproject [path]

Remove the placeholder starter runtime files from a <project>_composition
repository, leaving the service contract (.gitignore, .env.example) and the
repository itself intact so a real frontend/backend can replace the starter.

Arguments:
  path   path to the composition repository (default: nearest *_composition dir)

Options:
  --commit   commit the removal in the composition repository
  --dry-run  list what would be removed without deleting
"#;
    print!("{text}");
}

pub fn run_clear_starter_project(args: &[String]) -> Result<(), String> {
    if args.iter().any(|a| a == "help" || a == "--help" || a == "-h") {
        help_text();
        return Ok(());
    }
    let dry_run = args.iter().any(|a| a == "--dry-run");
    let commit = args.iter().any(|a| a == "--commit");
    let explicit_path = find_composition_arg(args);

    let cwd = util::current_dir();
    let composition_dir = match explicit_path {
        Some(p) => cwd.join(p),
        None => find_composition_dir(&cwd).ok_or(
            "No <project>_composition directory found. Pass the path or run inside the estate.",
        )?,
    };
    if !path_exists(&composition_dir) {
        return Err(format!(
            "Composition directory not found: {}",
            composition_dir.display()
        ));
    }

    let is_repo = composition_dir.join(".git").exists();
    if is_repo && !dry_run {
        match git_status_clean(&composition_dir) {
            Some(false) => {
                return Err(format!(
                    "{} has uncommitted changes to tracked files. \
Commit or stash them before clearing the starter project.",
                    composition_dir.display()
                ));
            }
            Some(true) => {}
            None => {}
        }
    }

    let mut present: Vec<&(&str, &str, bool)> = Vec::new();
    for entry in STARTER_PATHS {
        if path_exists(&composition_dir.join(entry.0)) {
            present.push(entry);
        }
    }

    let mut out = String::new();
    out.push_str(&format!(
        "\n{}  {}\n",
        util::bold("Eco starter project"),
        util::dim(&composition_dir.display().to_string())
    ));
    if present.is_empty() {
        out.push_str(&format!("  {} — no starter files present.\n\n", util::green("Nothing to clear")));
        print!("{out}");
        return Ok(());
    }

    for entry in &present {
        let verb = if dry_run { "would remove" } else { "removing" };
        out.push_str(&format!("  {verb}  {}\n", util::cyan(entry.0)));
    }
    if dry_run {
        out.push('\n');
        print!("{out}");
        return Ok(());
    }

    for entry in &present {
        let full = composition_dir.join(entry.0);
        let _ = std::fs::remove_dir_all(&full);
        let _ = std::fs::remove_file(&full);
    }

    for service_dir in ["frontend", "backend"] {
        let full = composition_dir.join(service_dir);
        if full.is_dir() {
            let remaining: usize = std::fs::read_dir(&full).map(|e| e.count()).unwrap_or(0);
            if remaining == 0 {
                let _ = std::fs::remove_dir_all(&full);
                out.push_str(&format!("  removing empty   {}\n", util::cyan(&format!("{service_dir}/"))));
            }
        }
    }

    if commit && is_repo {
        let _cwd = util::current_dir();
        let result = util::run_capture(
            "git",
            &["add".to_string(), "-A".to_string()],
            &composition_dir,
        )
        .map_err(|e| format!("git add: {e}"))?;
        if result.code != 0 {
            return Err("git add exited with non-zero".to_string());
        }
        let result = util::run_capture(
            "git",
            &["commit".to_string(), "-m".to_string(), "clear: remove starter project scaffold".to_string()],
            &composition_dir,
        )
        .map_err(|e| format!("git commit: {e}"))?;
        if result.code != 0 {
            return Err(format!("git commit failed: {}", result.stderr.trim()));
        }
        out.push_str(&format!("\n  {} removal in {}\n", util::green("Committed"), composition_dir.display()));
    } else if is_repo && !commit {
        out.push_str(&format!(
            "\n  {}\n",
            util::dim("Commit the removal with: git -C <composition> commit -m \"clear: remove starter scaffold\"")
        ));
    }
    let _ = cwd;
    out.push('\n');
    print!("{out}");
    Ok(())
}
